src/lib/CacheInstance.ts

```diff
@@ -143,17 +149,35 @@ export abstract class CacheInstance extends EventEmitter {
         lock = await this.lock(lockName, lockTtl * 1000);
         // check if the value has been populated while we were locking
         const cachedValue = await this.getValue(key);
+        if (cachedValue instanceof Error) {
+          throw cachedValue;
+        }
         if (cachedValue !== undefined) {
           return cachedValue;
         }
       }
 
       // fetch!
-      const fetchPromise = this.activeFetches[key] = fetchFunction();
+      let error: Error | undefined;
+      let result: any;
+      try {
+        const fetchPromise = this.activeFetches[key] = fetchFunction();
+        result = await fetchPromise;
+      } catch (err) {
+        error = err;
+      }
+
+      // cache! results: always, errors: only if satisfying user assertion
+      let errorToCache: Error | undefined;
+      if (error && shouldCacheError(error)) {
+        errorToCache = error;
+      }
+      if ((errorToCache || result) !== undefined) {
+        await this.setValue(key, errorToCache || result, ttl);
+      }
```

Comment:
I think it is simpler with two separate calls:

```suggestion
      if (error && shouldCacheError(error)) {
        await this.setValue(key, errorToCache, ttl);
      } else if (result !== undefined) {
        await this.setValue(key, result, ttl);
      }
```

Reply:
Hmmm I liked to express that caching errors vs. results took visibly the same path. But okay the "or" stuff is a little bit weird. 👍